diff --git a/3rdparty/wil b/3rdparty/wil
--- a/3rdparty/wil
+++ b/3rdparty/wil
@@ -1 +1 @@
-Subproject commit 3018875eaa4f8e3d86cf2da97fd577bae59f3aec
+Subproject commit 3018875eaa4f8e3d86cf2da97fd577bae59f3aec-dirty
diff --git a/pcsx2/CDVD/CDVD.cpp b/pcsx2/CDVD/CDVD.cpp
index 1bff89adc..dcf2998f9 100644
--- a/pcsx2/CDVD/CDVD.cpp
+++ b/pcsx2/CDVD/CDVD.cpp
@@ -81,10 +81,10 @@ static void CDVDSECTORREADY_INT(u32 eCycle)
 	if (psxRegs.interrupt & (1 << IopEvt_CdvdSectorReady))
 		return;
 
-	if (EmuConfig.Speedhacks.fastCDVD)
+	//if (EmuConfig.Speedhacks.fastCDVD)
 	{
 		if (eCycle < Cdvd_FullSeek_Cycles && eCycle > 1)
-			eCycle *= 0.5f;
+			eCycle *= 0.005f;
 	}
 
 	PSX_INT(IopEvt_CdvdSectorReady, eCycle);
@@ -94,10 +94,10 @@ static void CDVDREAD_INT(u32 eCycle)
 {
 	// Give it an arbitary FAST value. Good for ~5000kb/s in ULE when copying a file from CDVD to HDD
 	// Keep long seeks out though, as games may try to push dmas while seeking. (Tales of the Abyss)
-	if (EmuConfig.Speedhacks.fastCDVD)
+	//if (EmuConfig.Speedhacks.fastCDVD)
 	{
 		if (eCycle < Cdvd_FullSeek_Cycles && eCycle > 1)
-			eCycle *= 0.5f;
+			eCycle *= 0.005f;
 	}
 
 	PSX_INT(IopEvt_CdvdRead, eCycle);
